### DISCUSSION TITLE
Lesson 12: Defi Stablecoin. Part 13: Creating Redeem Function

### DISCUSSION THREAD

**Discussion by Hakob23** [Cyfrin/foundry-full-course-cu#1572] [Q&A] [answered]
Inside the ```redeemCollateral()``` function, the  ```_revertIfHealthFactorIsBroken()``` is called at the end, explained by fact that the health factor should be checked after the token transfer has happened. But the thing is that ```getAccountCollateralValue()``` is calculated by using the storage variables, which are being modified at the beginning of the ```redeemCollateral()``` function and thus, the result of  ```_revertIfHealthFactorIsBroken()``` function is independent from the ```IERC20.transfer()``` function. Then, why still it was called at the end? Am I missing something?
```
function redeemCollateral(address tokenCollateralAddress, uint256 amountCollateral)
external
nonReentrant
moreThanZero(amountCollateral){
      s_collateralDeposited[msg.sender][tokenCollateralAddress] -= amountCollateral;
      
      emit CollateralRedeemed(msg.sender, tokenCollateralAddress, amountCollateral)
      
      bool success = IERC20(tokenCollateralAddress).transfer(msg.sender, amountCollateral);
      if (!success){<
          revert DSCEngine__TransferFailed();
      }
      _revertIfHealthFactorIsBroken(msg.sender);
}

function getAccountCollateralValue(address user) public view returns (uint256 totalCollateralValueInUsd) {
        for (uint256 index = 0; index < s_collateralTokens.length; index++) {
            address token = s_collateralTokens[index];
            uint256 amount = s_collateralDeposited[user][token];
            totalCollateralValueInUsd += _getUsdValue(token, amount);
        }
        return totalCollateralValueInUsd;
    }
```

**Comment by EngrPips** [comment]
Hello @Hakob23 , I am not sure if i understand your question properly but from what i could understand my answer would be , we need to do the check `_revertIfHealthFactorIsBroken(msg.sender);` after we have already sent the user what they want to reedem so as to be sure we are doing the `healthfactor` calculation using the amount that is left after we have sent them what they requested to withdraw and if the `healthfactor` should pass that means they can withdraw what they intended to withdraw and still hold whatever position they are holding because they are still healthy. If i don't answer this as you expect please maybe clarify more to me so i can better understand.

**Reply by Hakob23** [reply]
Hey @EngrPips, thank you for the response. You see, according to the contract, the ```healthFactor``` is not measured by the ERC20 balances of the user but by the mappings of collaterals inside the contract itself, which means, that ```ERC20.transfer()``` function does not change the ```healthFactor``` at all.
```
function _getAccountInformation(address user)
        private
        view
        returns (uint256 totalDscMinted, uint256 collateralValueInUsd)
    {
        totalDscMinted = s_DSCMinted[user];
        collateralValueInUsd = getAccountCollateralValue(user);
    }

    function _healthFactor(address user) private view returns (uint256) {
        (uint256 totalDscMinted, uint256 collateralValueInUsd) = _getAccountInformation(user);
        return _calculateHealthFactor(totalDscMinted, collateralValueInUsd);
    }

    function _getUsdValue(address token, uint256 amount) private view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(s_priceFeeds[token]);
        (, int256 price,,,) = priceFeed.staleCheckLatestRoundData();
        // 1 ETH = 1000 USD
        // The returned value from Chainlink will be 1000 * 1e8
        // Most USD pairs have 8 decimals, so we will just pretend they all do
        // We want to have everything in terms of WEI, so we add 10 zeros at the end
        return ((uint256(price) * ADDITIONAL_FEED_PRECISION) * amount) / PRECISION;
    }

    function _calculateHealthFactor(uint256 totalDscMinted, uint256 collateralValueInUsd)
        internal
        pure
        returns (uint256)
    {
        if (totalDscMinted == 0) return type(uint256).max;
        uint256 collateralAdjustedForThreshold = (collateralValueInUsd * LIQUIDATION_THRESHOLD) / LIQUIDATION_PRECISION;
        return (collateralAdjustedForThreshold * PRECISION) / totalDscMinted;
    }

    function revertIfHealthFactorIsBroken(address user) internal view {
        uint256 userHealthFactor = _healthFactor(user);
        if (userHealthFactor < MIN_HEALTH_FACTOR) {
            revert DSCEngine__BreaksHealthFactor(userHealthFactor);
        }
    }
```
As you can see in the above mentioned functions, ```healthFactor``` is not depending on ERC20 balances of the user, but the mappings of collaterals that the user has, and the later is not modified by ```ERC20.transfer()``` function. Is there any other info that I could provide for the clarity?

**Comment by EngrPips** [comment]
Alright i think that makese, so lets walk through all of the process that would happen when a user try to redeem some of their collateral 

- A user requested to redeem 10 eth out of their 100 eth collateral so they call the function below 

```javascript 
function redeemCollateral(address _address_of_collateral_token, uint256 _amount_to_redeem)
        public
        more_than_zero(_amount_to_redeem)
        token_need_to_be_allowed(_address_of_collateral_token)
        nonReentrant
    {
        _redeem_collateral(msg.sender, msg.sender, _address_of_collateral_token, _amount_to_redeem);
        _revert_if_health_factor_is_broken(msg.sender);
    }
``` 

- The function above calls another internal function to facilitate the redemption , the internal function is the below 

```javascript
function _redeem_collateral(
        address _from_address,
        address _to_address,
        address _address_of_collateral_token,
        uint256 _amount_to_redeem
    ) internal {
        s_user_to_amount_deposited_by_user[_from_address][_address_of_collateral_token] -= _amount_to_redeem;
        emit collateral_redeemed(_from_address, _to_address, _address_of_collateral_token, _amount_to_redeem);
        (bool success) = IERC20(_address_of_collateral_token).transfer(_to_address, _amount_to_redeem);
        if (!success) revert DSCEngine__transfer_failed_when_user_try_to_redeem_collateral();
    }

```

The function above subtract the 10 ether from the 100 ether deposit that the user have with the protocol , then emit an event that tells the whole blockchain that the user has withdrawn 10 ether out of the 100 ether they had deposited with the protocol, then the protocol transfer 10 ether from the total balance of the protocol to the user that requested to withdraw 10 ether and their balance with the protocol is now 90 ether. if the transfer of the 10 ether didn't go through the protocol would revert with an error. 

- Now the protocol have successfully facilitate the redemption of the user collateral and we would go back inside the function the user called initially to continue the next line of action in that function, let's look at the function again below 

```javascript 
function redeemCollateral(address _address_of_collateral_token, uint256 _amount_to_redeem)
        public
        more_than_zero(_amount_to_redeem)
        token_need_to_be_allowed(_address_of_collateral_token)
        nonReentrant
    {
        _redeem_collateral(msg.sender, msg.sender, _address_of_collateral_token, _amount_to_redeem);
        _revert_if_health_factor_is_broken(msg.sender);
    }
``` 
So we have successfully redeemed the user collateral and now the next line of action is to check that the user health factor is not broken as a result of this redemption.  now we call the second line `_revert_if_health_factor_is_broken(msg.sender);` in the function to check the user health factor and ensure that the health factor is still intact otherwise we would revert the whole redemption. The second line is an internal function is shown below 

```javascript
function _revert_if_health_factor_is_broken(address _user) internal view {
        uint256 user_health_factor = _health_factor(_user);
        if (user_health_factor < MINIMUM_HEALTH_FACTOR) revert DSCEngine__health_factor_is_broken();
    }
```
The above function collect the address of the user who just redeemed their collateral and pass it to an internal function `_health_factor(_user);` which would get the health factor of the user, let's see what the internal function does below.

```javascript
function _health_factor(address _user) internal view returns (uint256 health_factor) {
        (uint256 total_DSC_minted, uint256 collateral_value_in_USD) = _get_account_information(_user);
        // console2.log("the total collateral value in usd is: ", collateral_value_in_USD);
        // console2.log("the total DSC minted is: ", total_DSC_minted);
        if (total_DSC_minted == 0) return type(uint256).max;
        uint256 collateral_amount_for_threshold =
            (collateral_value_in_USD * LIQUIDATION_THRESHOLD) / LIQUIDATION_THRESHOLD_DIVIDER;
        // console2.log("the collateral amount for threshold is: ", collateral_amount_for_threshold);
        health_factor = collateral_amount_for_threshold  / total_DSC_minted;
        // console2.log("the calculated health factor is: ", health_factor);
        return health_factor;
    }
```
The above function also pass the user address to an internal function `_get_account_information(_user);`  which would then return the total `DSC` that user have minted and the total collateral that the user have with the protocol, lets see what the internal function looks like below.

```javascript
function _get_account_information(address _user)
        internal
        view
        returns (uint256 total_DSC_minted, uint256 collateral_value_in_USD)
    {
        total_DSC_minted = s_user_to_amount_of_DSC_minted[_user];
        collateral_value_in_USD = get_account_collateral_value_in_USD(_user);

        return (total_DSC_minted, collateral_value_in_USD);
    }

```
The above function reads the storage of the contract to get the `total_DSC_minted` minted by the user and then pass the address to an internal function `get_account_collateral_value_in_USD`  which would help with getting the total collateral that the user have with the protocol. let's look what the internal function looks like below 

```javascript
function get_account_collateral_value_in_USD(address _user)
        public
        view
        returns (uint256 total_collateral_value_in_USD)
    {
        for (uint256 i = 0; i < s_collateral_tokens.length; i++) {
            address token = s_collateral_tokens[i];
            uint256 amount = s_user_to_amount_deposited_by_user[_user][token];
            total_collateral_value_in_USD += get_USD_value(token, amount);
        }

        return total_collateral_value_in_USD;
    }
```
The above function collects the address of the user which we want to check their health factor and run a loop that is the length of the `s_collateral_tokens.length` and for each of the collateral token supported on the protocol it gets the total amount that the user has deposited for each of them and then get the USD value for each of the amount of collateral token the user have deposited. 

**IT IS VERY IMPORTANT TO REMEMBER THAT WE HAVE DEDUCTED WHAT THE USER REDEEMED FROM THE COLLATERAL THEY REDEEMED FROM SO THAT WOULD BE ACCOUNTED FOR IN THIS FOR LOOP WHERE THE TOTAL COLLATERAL OF THE USER IS CALCULATED`**

so the above function gets the total USD value for all of the collateral deposited by the user, returns it to the function below  

```javascript
function _get_account_information(address _user)
        internal
        view
        returns (uint256 total_DSC_minted, uint256 collateral_value_in_USD)
    {
        total_DSC_minted = s_user_to_amount_of_DSC_minted[_user];
        collateral_value_in_USD = get_account_collateral_value_in_USD(_user);

        return (total_DSC_minted, collateral_value_in_USD);
    }

```
Then this function above returns both the `total_DSC_minted` and `collateral_value_in_USD` to the function below 

```javascript
function _health_factor(address _user) internal view returns (uint256 health_factor) {
        (uint256 total_DSC_minted, uint256 collateral_value_in_USD) = _get_account_information(_user);
        // console2.log("the total collateral value in usd is: ", collateral_value_in_USD);
        // console2.log("the total DSC minted is: ", total_DSC_minted);
        if (total_DSC_minted == 0) return type(uint256).max;
        uint256 collateral_amount_for_threshold =
            (collateral_value_in_USD * LIQUIDATION_THRESHOLD) / LIQUIDATION_THRESHOLD_DIVIDER;
        // console2.log("the collateral amount for threshold is: ", collateral_amount_for_threshold);
        health_factor = collateral_amount_for_threshold  / total_DSC_minted;
        // console2.log("the calculated health factor is: ", health_factor);
        return health_factor;
    }
```
The function above uses both the `total_DSC_minted` and `collateral_value_in_USD` to calculate the health factor of the user and return the health factor of the user to the function below.

```javascript
function _revert_if_health_factor_is_broken(address _user) internal view {
        uint256 user_health_factor = _health_factor(_user);
        if (user_health_factor < MINIMUM_HEALTH_FACTOR) revert DSCEngine__health_factor_is_broken();
    }
```

The function above gets the health factor of the user then compare it to the `MINIMUM_HEALTH_FACTOR` and if it turns out that the health factor of the user is lower than the `MINIMUM_HEALTH_FACTOR` the whole redemption process of the user would be reverted like it never happened.

I really hope i am able to make you understand, if anything still seem confusing please let me know

**Comment by EngrPips** [comment]
And this your statement `and the later is not modified by ERC20.transfer() function. Is there any other info that I could provide for the clarity?` is actually correct but we modified  the mapping of the collateral the user have with the protocol  in the first line of  this function below, this line `s_user_to_amount_deposited_by_user[_from_address][_address_of_collateral_token] -= _amount_to_redeem;`. Yes `ERC20.transfer()` doesn't modify the collateral the user have with the protocol but we did that manually by ourselves.

```
function _redeem_collateral(
        address _from_address,
        address _to_address,
        address _address_of_collateral_token,
        uint256 _amount_to_redeem
    ) internal {
        s_user_to_amount_deposited_by_user[_from_address][_address_of_collateral_token] -= _amount_to_redeem;
        emit collateral_redeemed(_from_address, _to_address, _address_of_collateral_token, _amount_to_redeem);
        (bool success) = IERC20(_address_of_collateral_token).transfer(_to_address, _amount_to_redeem);
        if (!success) revert DSCEngine__transfer_failed_when_user_try_to_redeem_collateral();
    }
```

**Reply by Hakob23** [reply]
Okay, now I see. The final version of the function is significantly different from its initial version(minute 5:14). And my question is relevant to the later one. I will bring that version of the function here so you can see.
```
function redeemCollateral(address tokenCollateralAddress, uint256 amountCollateral)
external
nonReentrant
moreThanZero(amountCollateral){
s_collateralDeposited[msg.sender][tokenCollateralAddress] -= amountCollateral;

emit CollateralRedeemed(msg.sender, tokenCollateralAddress, amountCollateral)

bool success = IERC20(tokenCollateralAddress).transfer(msg.sender, amountCollateral);
if (!success){
    revert DSCEngine__TransferFailed();
}
_revertIfHealthFactorIsBroken(msg.sender);
}
```
why could not we just put the ```_revertIfHealthFactorIsBroken(msg.sender);``` function above the ```IERC20.transfer()``` if the result for ```_revertIfHealthFactorIsBroken(msg.sender);``` is going to be the same regardless ```IERC20.transfer()``` and by that we could preserve CEI structure? Hope this makes more sense now. Thank you for your time and previous answer.

**Reply by EngrPips** [reply, marked as answer]
I see , thats just design choice. you could do it that way too and nothing changes.

**Reply by Hakob23** [reply]
Alright, thank you :)

**Reply by EngrPips** [reply]
Anytime friend, and moreover we already added  Openzeppelin `nonReentrant` modifier to the function so it can not be re-entered either we follow `CEI` or not

**Reply by Hakob23** [reply]
Got it, thank you :)

**Reply by EngrPips** [reply]
You are welcome friend !
